parser-typechecker/src/Unison/Hashing/V2/Hashable.hs

```diff
@@ -0,0 +1,139 @@
+module Unison.Hashing.V2.Hashable where
+
+import qualified Crypto.Hash as CH
+import qualified Data.ByteArray as BA
+import qualified Data.ByteString as B
+import Data.ByteString.Builder (doubleBE, int64BE, toLazyByteString, word64BE)
+import qualified Data.ByteString.Lazy as BL
+import qualified Data.Map as Map
+import qualified Data.Set as Set
+import U.Util.Hash (Hash)
+import qualified U.Util.Hash as Hash
+import Unison.Prelude
+import Unison.Util.Relation (Relation)
+import qualified Unison.Util.Relation as Relation
+import Unison.Util.Relation3 (Relation3)
+import qualified Unison.Util.Relation3 as Relation3
+import Unison.Util.Relation4 (Relation4)
+import qualified Unison.Util.Relation4 as Relation4
+
+data Token h
+  = Tag !Word8
+  | Bytes !ByteString
+  | Int !Int64
+  | Text !Text
+  | Double !Double
+  | Hashed !h
+  | Nat !Word64
+
+class Accumulate h where
+  accumulate :: [Token h] -> h
+  fromBytes :: ByteString -> h
+  toBytes :: h -> ByteString
+
+accumulateToken :: (Accumulate h, Hashable t) => t -> Token h
+accumulateToken = Hashed . accumulate'
+
+accumulate' :: (Accumulate h, Hashable t) => t -> h
+accumulate' = accumulate . (hashVersion :) . tokens
+  where
+    hashVersion = Tag 2
+
+class Hashable t where
+  tokens :: Accumulate h => t -> [Token h]
+
+instance Hashable a => Hashable [a] where
+  tokens = map accumulateToken
+
+instance (Hashable a, Hashable b) => Hashable (a, b) where
+  tokens (a, b) = [accumulateToken a, accumulateToken b]
+
+instance (Hashable a) => Hashable (Set.Set a) where
+  tokens = tokens . Set.toList
+
+instance (Hashable k, Hashable v) => Hashable (Map.Map k v) where
+  tokens = tokens . Map.toList
+
+instance (Hashable a, Hashable b) => Hashable (Relation a b) where
+  tokens = tokens . Relation.toList
+
+instance (Hashable d1, Hashable d2, Hashable d3) => Hashable (Relation3 d1 d2 d3) where
+  tokens s = [accumulateToken $ Relation3.toNestedList s]
+
+instance (Hashable d1, Hashable d2, Hashable d3, Hashable d4) => Hashable (Relation4 d1 d2 d3 d4) where
+  tokens s = [accumulateToken $ Relation4.toNestedList s]
+
+class Functor f => Hashable1 f where
```

Comment:
I know you're just moving this, but does this actually need the Functor constraint? Seems like it doesn't depend on it, and the implementer of the instance should probably be able to decide their own constraints right?

I don't mind if it stays, mostly just asking out of curiousity.

Reply:
Seems like an oversight and that it should go.